qoffer-domain/src/test/groovy/life/qbic/portal/portlet/offers/OfferSpec.groovy

```diff
@@ -0,0 +1,154 @@
+package life.qbic.portal.portlet.offers
+
+import life.qbic.datamodel.dtos.business.Affiliation
+import life.qbic.datamodel.dtos.business.AffiliationCategory
+import life.qbic.datamodel.dtos.business.Customer
+import life.qbic.datamodel.dtos.business.ProductItem
+import life.qbic.datamodel.dtos.business.ProjectManager
+import life.qbic.datamodel.dtos.business.services.DataStorage
+import life.qbic.datamodel.dtos.business.services.PrimaryAnalysis
+import life.qbic.datamodel.dtos.business.services.ProductUnit
+import life.qbic.datamodel.dtos.business.services.ProjectManagement
+import life.qbic.portal.portlet.offers.Offer
+import spock.lang.Shared
+import spock.lang.Specification
+
+/**
+ * Test for the business rules in offer calculus at QBiC.
+ *
+ * @since 1.0.0
+ */
+class OfferSpec extends Specification {
+
+    @Shared
+    Affiliation externalAcademicAffiliation
+    @Shared
+    Affiliation externalAffiliation
+    @Shared
+    Affiliation internalAffiliation
+    @Shared
+    Customer customerWithAllAffiliations
+    @Shared
+    ProjectManager projectManager
+
+    def setup() {
+        internalAffiliation = new Affiliation.Builder("Uni Tübingen", "Auf der " +
+                "Morgenstelle 10", "72076", "Tuebingen").category(AffiliationCategory.INTERNAL).build()
+        externalAcademicAffiliation = new Affiliation.Builder("Uni Frankfurt",
+                "Irgendwo im Nirgendwo 20", "12345", "Frankfurt").category(AffiliationCategory
+                .EXTERNAL_ACADEMIC)
+                .build()
+        externalAffiliation = new Affiliation.Builder("Company Frankfurt",
+                "Irgendwo " +
+                        "im Nirgendwo 20", "12345", "Frankfurt").category(AffiliationCategory.EXTERNAL)
+                .build()
+        customerWithAllAffiliations = new Customer.Builder("Max", "Mustermann", "max" +
+                ".mustermann@qbic.uni-tuebingen.de").affiliations([internalAffiliation, externalAffiliation]).build()
+        projectManager  = new ProjectManager.Builder("Maxime", "Musterfrau", "max" +
+                ".musterfrau@qbic.uni-tuebingen.de").build()
+    }
+
+    def "A customer with an internal affiliation shall pay no overheads"() {
+        given:
+        List<ProductItem> items = [
+                new ProductItem(2, new PrimaryAnalysis("Basic RNAsq", "Just an" +
+                        " example", 1.0, ProductUnit.PER_SAMPLE)),
+                new ProductItem(1, new ProjectManagement("Basic Management",
+                        "Just an example", 10.0, ProductUnit.PER_DATASET))
```

Comment:
why is this an improvement?

Reply:
Because you can easily test a few edge cases ( e.g. what happens when the quantity is 0 or the price of a product is 0)
>         where:
        itemCost1 | amount1 | itemCost2 | amount2
        10 | 1 | 10 | 2
        0 | 1 | 10 | 2
        10 | 0 | 10 | 2
        10 | 1 | 10 | 0
        10 | 1 | 0 | 2

Reply:
And you can answer the question if this also holds true for different inputs.

Reply:
yes, my point is, we have not even decided, if 0 values are permitted here at all on a business level. It does not make sense to have a service with 0 unit price or an item with 0 amount.

Reply:
yes I get that 😄 and I fully agree. It was just meant as an example of when multiple inputs could be useful.